packages/agent/lib/agent.ts

```diff
@@ -0,0 +1,266 @@
+    }
+
+    // node has queued an HTTP message to be sent already, so we need to regenerate the
+    // queued message with the new path and headers
+    // https://github.com/TooTallNate/node-http-proxy-agent/blob/master/index.js#L93
+    regenerateRequestHead(req)
+
+    options.port = Number(proxy.port || 80)
+    options.host = proxy.hostname || 'localhost'
+    delete options.path // so the underlying net.connect doesn't default to IPC
+
+    if (proxy.protocol === 'https:') {
+      // gonna have to use the https module to reach the proxy, even though this is an http req
+      req.agent = this.httpsAgent
+
+      return this.httpsAgent.addRequest(req, options)
+    }
+
+    super.createSocket(req, options, cb)
+  }
+}
+
+class HttpsAgent extends https.Agent {
+  constructor (opts: https.AgentOptions = {}) {
+    opts.keepAlive = true
+    super(opts)
+  }
+
+  createConnection (options: http.RequestOptions, cb: http.SocketCallback) {
+    if (process.env.HTTPS_PROXY) {
+      const proxy = getProxyForUrl(options.href)
+
+      if (typeof proxy === "string") {
+        options.proxy = <string>proxy
+
+        return this.createProxiedConnection(<RequestOptionsWithProxy>options, cb)
+      }
+    }
+
+    // @ts-ignore
+    cb(null, super.createConnection(options))
+  }
+
+  createProxiedConnection (options: RequestOptionsWithProxy, cb: http.SocketCallback) {
+    // heavily inspired by
+    // https://github.com/mknj/node-keepalive-proxy-agent/blob/master/index.js
+    debug(`Creating proxied socket for ${options.href} through ${options.proxy}`)
+
+    const proxy = url.parse(options.proxy)
+    const port = options.uri.port || 443
+    const hostname = options.uri.hostname
+
+    const proxySocket = createProxySock(proxy)
+
+    const onError = (err: Error) => {
+      proxySocket.destroy()
+      cb(err, undefined)
+    }
+
+    let buffer = ''
+
+    const onData = (data: Buffer) => {
+      debug(`Proxy socket for ${options.href} established`)
+
+      buffer += data.toString()
+
+      if (!_.includes(buffer, '\r\n\r\n')) {
+        // haven't received end of headers yet, keep buffering
+        proxySocket.once('data', onData)
+        return
+      }
+
+      proxySocket.removeListener('error', onError)
+      // read status code from proxy's response
+      const matches = buffer.match(/^HTTP\/1.1 (\d*)/)
+
+      if (!matches || matches[1] !== '200') {
+        return onError(new Error(`Error establishing proxy connection: ${matches ? matches[0] : buffer}`))
+      }
+
+      if (options._agentKey) {
+        // https.Agent will upgrade and reuse this socket now
+        options.socket = proxySocket
+        options.servername = hostname
+        return cb(undefined, super.createConnection(options, undefined))
+      }
+
+      cb(undefined, proxySocket)
+    }
+
+    proxySocket.once('error', onError)
+    proxySocket.once('data', onData)
+
+    let connectReq = `CONNECT ${hostname}:${port} HTTP/1.1\r\n`
```

Comment:
I would build up an array here rather than a string and then do a `.join('\r\n')` at the end... much cleaner